transport/src/main/java/io/netty/bootstrap/ChannelInitializerExtension.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2023 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.bootstrap;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ServerChannel;
+
+/**
+ * A channel initializer extension make it possible to enforce rules and apply modifications across multiple,
+ * disconnected uses of Netty within the same JVM process.
+ * <p>
+ * For instance, application-level firewall rules can be injected into all uses of Netty within an application,
+ * without making changes to such uses that are otherwise outside the purview of the application code,
+ * such as 3rd-party libraries.
+ * <p>
+ * All channel initializer extensions that are available on the classpath will be
+ * {@linkplain java.util.ServiceLoader#load(Class) service-loaded} and used by all {@link AbstractBootstrap} subclasses,
+ * unless the extensions are explicitly disabled on the given bootstrap instance with a call to
+ * {@link AbstractBootstrap#disableChannelInitializerExtensions()}.
+ */
+public abstract class ChannelInitializerExtension {
+    /**
+     * Inspect the information in the given {@link ApplicableInfo} object, and determine if this extension is applicable
+     * for the given context.
+     * <p>
+     * If the extension is not applicable, then it won't be called.
+     * <p>
+     * This method may be called multiple times with different parameters for different contexts.
+     *
+     * @param info An object carrying information about the context of where and how this extension would be used.
+     * @return {@code true} if this extension is interested in getting called in the given context,
+     * otherwise {@code false}.
+     */
+    public abstract boolean isApplicable(ApplicableInfo info);
```

Comment:
Should we default to `true` or do we expect users always check the bootstrap type inside `info`?

Reply:
We can probably get away with a default.